### PR TITLE
Add org.zdoom.GZDoom

### PR DESCRIPTION
**Pull request by Eonfge**
We already have a few packages (Freedom, Blade of Agony) that make use of the GZDoom engine, but we don't actually have the engine 'stand-alone' to use with Doom, Heretic and the rest.

By adding GZDoom to the growing list of Flatpaks, we instantly add whole library of compatible titles to the list.

### CODE REVIEW

**Comment by Eonfge** [comment]
bot, build

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

File: org.zdoom.GZDoom.desktop [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[Desktop Entry]
+Categories=Games;
+Comment=Multi-game launcher for Doom era games
+Exec=gzdoom
+Icon=gzdoom.png
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.desktop]:
```suggestion
Icon=org.zdoom.GZDoom
```

File: org.zdoom.GZDoom.yaml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    path: org.zdoom.GZDoom.desktop
+  - type: file
+    path: org.zdoom.GZDoom.appdata.xml
+  - type: file
+    path: gzdoom.png
+  build-commands:
+  - install -D gzdoom.sh /app/bin/gzdoom.sh
+  - install -Dm 644 org.zdoom.GZDoom.desktop -t /app/share/applications
+  - install -Dm 644 org.zdoom.GZDoom.appdata.xml -t /app/share/appdata
+  - install -Dm 644 gzdoom.png /app/share/icons/hicolor/48x48/apps/gzdoom.png
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.yaml]: Just install the icons with the correct name here.

File: org.zdoom.GZDoom.yaml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  - type: file
+    path: org.zdoom.GZDoom.desktop
+  - type: file
+    path: org.zdoom.GZDoom.appdata.xml
+  - type: file
+    path: gzdoom.png
+  build-commands:
+  - install -D gzdoom.sh /app/bin/gzdoom.sh
+  - install -Dm 644 org.zdoom.GZDoom.desktop -t /app/share/applications
+  - install -Dm 644 org.zdoom.GZDoom.appdata.xml -t /app/share/appdata
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.yaml]:
```suggestion
  - install -Dm 644 org.zdoom.GZDoom.appdata.xml -t /app/share/metainfo
```
the `$datadir/appdata` shouldn't be used anymore.

File: org.zdoom.GZDoom.yaml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+# Screenshots? 
+# Sceenshots go by default in ~/.config/
+#- --filesystem=xdg-pictures
+
+# Based on GZDoom
+# We access the original ~/.config/gzdoom because
+# that is where users should put their iwads
+- --env=DOOMWADDIR=/app/share/games/doom
+- --filesystem=~/.config/gzdoom
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.yaml]: Not sure if was this discussed for other games, can't we have it use the xdg-data directory?

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

File: org.zdoom.GZDoom.yaml (hunk 4 above)
Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: Problem is, that GZDoom doesn't actually have any game dta files, so users are asked to provide their own. I'm not sure what is the best way in letting users to add their own proprietary data. The current implementation of GZDoom does ask the user to add them to ~/.config/GZDoom, so I decided to stick with that for now.

I did play around with a _persist_ but it only made it more confusing

File: org.zdoom.GZDoom.desktop [hunk 5]
```diff
@@ -2,7 +2,7 @@
 Categories=Games;
 Comment=Multi-game launcher for Doom era games
 Exec=gzdoom
-Icon=gzdoom.png
+Icon=org.zdoom.GZDoom.png
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.desktop]: No need for the extension.

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
The image gallery btw:
https://imgur.com/a/46I7Wwf

I picked some iconic locations to screenshot, but I did try to stay away from images depicting any spoilers or violence.

**Comment by Eonfge** [comment]
_cough, bot, cough_
bot, build org.zdoom.GZDoom

File: org.zdoom.GZDoom.yaml (hunk 4 above)
Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: For clarity, technically it's completely possible. It's more about consistency to the front-end user. This is the most likely place where they already have all their game data.

File: org.zdoom.GZDoom.yaml [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+finish-args:
+- --device=dri
+- --socket=wayland
+- --socket=fallback-x11
+- --socket=x11
+- --share=ipc
+- --socket=pulseaudio
+
+# Where would wads be?
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.yaml]: Please follow what other games do and remove those filesystem access. 
https://github.com/flathub/com.moddb.TotalChaos/blob/master/com.moddb.TotalChaos.yaml#L74

File: org.zdoom.GZDoom.yaml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+- --filesystem=xdg-desktop:ro
+- --filesystem=xdg-documents:ro
+- --filesystem=xdg-download:ro
+
+# Screenshots?
```
Comment (bilelmoussaoui) [inline org.zdoom.GZDoom.yaml]: Not needed.

File: org.zdoom.GZDoom.yaml (hunk 6 above)
Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: I removed the :ro directories. Looking back at our conversation surrounding ~/.config/ they are indeed not required.

**Comment by LeandroStanger** [comment]
bot, build org.zdoom.GZDoom

File: org.zdoom.GZDoom.yaml (hunk 7 above)
Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: Removed commentary

File: org.zdoom.GZDoom.yaml (hunk 6 above)
Comment (bilelmoussaoui) [reply org.zdoom.GZDoom.yaml]: The other games do download & install the wads required to play the game in `/app/share/games/doom` which is the reason of setting this env variable `--env=DOOMWADDIR=/app/share/games/doom`
The game must work out of the box without asking the user do anything to get it working.

Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: GZDoom also requires some wads which are stored in _/app/share/games/doom_.  during compilation, the system creates 4 .pk3 files which have essential compatibility data. As such, the env= parameter is required.

Comment (bilelmoussaoui) [reply org.zdoom.GZDoom.yaml]: Right. I'm not sure how we can handle this as exposing a folder where we would expect the user to install their extra wads files is not the best way of handling this. 
@barthalion What do you suggest doing in such case?

Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: It's indeed a bit of a conflict. We package some additional game data, which is a hard dependency, but we must also allow users to add aditional wads.

Right now it is like this: 
- ~/.config/gzdoom => user game data
- /app/share/games/doom/ => application game data

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by Eonfge** [comment]
@bilelmoussaoui I've reconsidered the situation. 

While it would still be workable in this scenario to 'claim' ~/.config/gzdoom for ourselves, it would be better for cross app support to not do that. See, SLADE for example also uses GZDoom and we don't want to force our game settings onto the editor's settings. 

Thus, I've made a git patch to change the user instruction set, telling users to add their game data to ~/.var/[...] instead.

File: org.zdoom.GZDoom.yaml [hunk 8]
```diff
@@ -0,0 +1,100 @@
+app-id: org.zdoom.GZDoom
+runtime: org.freedesktop.Platform
+sdk: org.freedesktop.Sdk
+runtime-version: "18.08"
```
Comment (LeandroStanger) [inline org.zdoom.GZDoom.yaml]: Update runtime to 19.08

Comment (Eonfge) [reply org.zdoom.GZDoom.yaml]: Took some time (As in, I previously looked into this problem three times) but I found out how to update the runtime without the GCC linker crashing.

**Comment by Eonfge** [comment]
bot, build org.zdoom.GZDoom

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/org.zdoom.GZDoom Read https://github.com/flathub/flathub/wiki/App-Maintenance for more on how Flathub works for app builders and accept the GitHub request you'll soon be sent!

**Pull request closed without merging**